src/metaxy/models/filter_expression.py

```diff
@@ -0,0 +1,314 @@
+"""SQL filter string parsing into Narwhals expressions.
+
+This module exposes utilities to parse SQL WHERE-like strings into Narwhals
+``Expr`` objects. The primary entry point is :func:`parse_filter_string`, which
+returns a Narwhals expression that can be fed directly into
+``LazyFrame.filter`` (works across all Narwhals backends).
+"""
+
+from __future__ import annotations
+
+from typing import Any, NamedTuple
+
+import narwhals as nw
+import sqlglot
+from pydantic import field_serializer, model_validator
+from sqlglot import exp
+from sqlglot.errors import ParseError
+
+from metaxy.models.bases import FrozenBaseModel
+
+LiteralValue = bool | int | float | str | None
+
+
+class FilterParseError(ValueError):
+    """Raised when a filter string cannot be parsed into a supported expression."""
+
+
+class OperandInfo(NamedTuple):
+    expr: nw.Expr
+    is_literal: bool
+    literal_value: LiteralValue
+    is_column: bool
+
+
+class NarwhalsFilter(FrozenBaseModel):
+    """Pydantic model for serializable Narwhals filter expressions."""
+
+    expression: sqlglot.exp.Expression
+    source: str | None = None
+
+    model_config = {
+        "arbitrary_types_allowed": True,
+        "extra": "forbid",
+        "frozen": True,
+    }
+
+    @model_validator(mode="before")
+    @classmethod
+    def _parse_expression_from_string(cls, data: Any) -> Any:
+        if isinstance(data, str):
+            expression = _parse_to_sqlglot_expression(data)
+            return {"expression": expression, "source": data}
+        return data
+
+    @field_serializer("expression")
+    def _serialize_expression(self, expression: sqlglot.exp.Expression) -> str:
+        return expression.sql()
+
+    def to_expr(self) -> nw.Expr:
+        """Convert the stored expression into a Narwhals ``Expr``."""
+        return _expression_to_narwhals(self.expression)
+
+
+def parse_filter_string(filter_string: str) -> nw.Expr:
+    """Parse a SQL WHERE-like string into a Narwhals expression.
+
+    The parser uses SQLGlot under the hood and understands SQL WHERE
+    clauses composed of comparison operators, logical operators, parentheses,
+    dotted identifiers, and literal values (strings, numbers, booleans, ``NULL``).
+    It returns a backend-agnostic Narwhals expression that can be fed directly
+    into ``LazyFrame.filter`` for any Narwhals-compatible backend (Polars,
+    pandas, PyArrow, DuckDB, ...).
+
+    Examples
```

Comment:
Looks like the format here isn't mkdocs

Are you looking at docs locally? 

`just docs-serve`

Reply:
You have to add it, see below

Reply:
fixed now